beacon-chain/rpc/beaconv1/validator.go

```diff
@@ -2,27 +2,159 @@ package beaconv1
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get state: %v", err)
+	}
+
+	valContainers, err := valContainersByRequestIds(state, req.Id)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get validator: %v", err)
+	}
+	valBalances := make([]*ethpb.ValidatorBalance, len(valContainers))
+	for i := 0; i < len(valContainers); i++ {
+		valBalances[i] = &ethpb.ValidatorBalance{
+			Index:   valContainers[i].Index,
+			Balance: valContainers[i].Balance,
+		}
+	}
+	return &ethpb.ValidatorBalancesResponse{Data: valBalances}, nil
 }
 
 // ListCommittees retrieves the committees for the given state at the given epoch.
+// If the requested slot and index are defined, only those committees are returned.
 func (bs *Server) ListCommittees(ctx context.Context, req *ethpb.StateCommitteesRequest) (*ethpb.StateCommitteesResponse, error) {
-	return nil, errors.New("unimplemented")
+	state, err := bs.StateFetcher.State(ctx, req.StateId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get state: %v", err)
+	}
+
+	epoch := helpers.SlotToEpoch(state.Slot())
+	if req.Epoch != nil {
+		epoch = *req.Epoch
+	}
+	activeCount, err := helpers.ActiveValidatorCount(state, epoch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get active validator count: %v", err)
+	}
+
+	startSlot, err := helpers.StartSlot(epoch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get epoch start slot: %v", err)
+	}
+	endSlot, err := helpers.EndSlot(epoch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get epoch end slot: %v", err)
+	}
+	committeesPerSlot := helpers.SlotCommitteeCount(activeCount)
+	committees := make([]*ethpb.Committee, 0)
+	for slot := startSlot; slot <= endSlot; slot++ {
+		if req.Slot != nil && slot != *req.Slot {
+			continue
+		}
+		for index := types.CommitteeIndex(0); index < types.CommitteeIndex(committeesPerSlot); index++ {
+			if req.Index != nil && index != *req.Index {
+				continue
+			}
+			committee, err := helpers.BeaconCommitteeFromState(state, slot, index)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "Could not get committee: %v", err)
+			}
+			committeeContainer := &ethpb.Committee{
+				Index:      index,
+				Slot:       slot,
+				Validators: committee,
+			}
+			committees = append(committees, committeeContainer)
+		}
+	}
+	return &ethpb.StateCommitteesResponse{Data: committees}, nil
+}
+
+// This function returns the validator object based on the passed in ID. The validator ID could be its public key,
+// or its index and is usually retrieved from the API request query.
+func valContainersByRequestIds(state iface.BeaconState, validatorIds [][]byte) ([]*ethpb.ValidatorContainer, error) {
+	allValidators := state.Validators()
+	allBalances := state.Balances()
+	var valContainers []*ethpb.ValidatorContainer
+	if len(validatorIds) == 0 {
```

Comment:
Please also write tests for this condition.

Reply:
Added!